File: 0077-combinations/0077-combinations.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        // Initialize the result list to store all combinations
        List<List<Integer>> list = new ArrayList<>();
        
        // Start backtracking from the first element
        backtrack(list, n, k, new ArrayList<>(), 1);
        return list;
    }

    private void backtrack(List<List<Integer>> list, int n, int k, List<Integer> temp, int start) {
        // Base case: If k is zero, add the current combination to the result list
        if (k == 0) {
            list.add(new ArrayList<>(temp));
            return;
        }

        // Iterate from start to the maximum possible value for the combination
        for (int i = start; i <= n - k + 1; i++) {
            // Add the current number to the temporary list
            temp.add(i);

            // Recursively backtrack with the remaining numbers and one less element to pick
            backtrack(list, n, k - 1, temp, i + 1);

            // Remove the last added number to backtrack and try the next number
            temp.remove(temp.size() - 1);
        }
    }
}

File: 0077-combinations/0077-combinations.py
from typing import List

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        # List to store all combinations
        combs = []
        
        # List to store the current combination being constructed
        subs = []
        
        # Helper function for backtracking
        def backtrack(i, subs, combs, k, n):
            # If the current combination is of the desired length, add it to the result list
            if len(subs) == k:
                combs.append(subs.copy())  # Use copy to avoid reference issues
                return
            
            # If the current index exceeds n, stop the recursion
            if i > n:
                return
            
            # Include the current element and proceed
            subs.append(i)
            backtrack(i + 1, subs, combs, k, n)
            
            # Exclude the current element (backtrack) and proceed
            subs.pop()
            backtrack(i + 1, subs, combs, k, n)
        
        # Start the backtracking process from 1
        backtrack(1, subs, combs, k, n)
        
        # Return the list of all combinations
        return combs
